src/helpers.c

```diff
@@ -149,130 +149,108 @@ void concat_str_array(char **output, char *const argv[])
+		if (close(pipefd[0]) == -1) {
+			LOG(NULL, "Could not close read end of pipe file descriptor", "%d", pipefd[0]);
+			assert(0);
 		}
+
+		int lhsresult = wait_process_terminate(lhspid);
+		int rhsresult = wait_process_terminate(rhspid);
+		exit(rhsresult != EXIT_SUCCESS ? rhsresult : lhsresult);
 	}
+	return wait_process_terminate(monitorpid);
+}
 
-	return status;
+int system_argv(char *const argv[])
+{
+	return system_argv_fd(-1, -1, -1, argv);
 }
 
-int system_argv_fd(char *const argv[], int newstdin, int newstdout, int newstderr)
+int system_argv_fd(int newstdinfd, int newstdoutfd, int newstderrfd, char *const cmd[])
 {
-	int child_exit_status;
-	pid_t pid;
-	int status = -1;
-
-	pid = fork();
-
-	if (pid == 0) { /* child */
-		if (newstdin >= 0) {
-			if (dup2(newstdin, STDIN_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stdin", "");
-				assert(0);
-			}
-			close(newstdin);
-		}
-		if (newstdout >= 0) {
-			if (dup2(newstdout, STDOUT_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stdout", "");
-				assert(0);
-			}
-			close(newstdout);
-		}
-		if (newstderr >= 0) {
-			if (dup2(newstderr, STDERR_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stderr", "");
-				assert(0);
-			}
-			close(newstderr);
-		}
+	pid_t cmdpid = system_argv_fd_nowait(newstdinfd, newstdoutfd, newstderrfd, cmd);
+	return wait_process_terminate(cmdpid);
+}
 
-		execvp(*argv, argv);
-		LOG(NULL, "This line must not be reached", "");
+pid_t system_argv_fd_nowait(int newstdinfd, int newstdoutfd, int newstderrfd, char *const cmd[])
+{
+	pid_t cmdpid = fork();
+	if (cmdpid == -1) {
+		LOG(NULL, "Failed to fork to execute command", "%s", cmd[0]);
 		assert(0);
-	} else if (pid < 0) {
-		LOG(NULL, "Failed to fork a child process", "");
+	} else if (cmdpid == 0) {
+		set_std_fd(newstdinfd, newstdoutfd, newstderrfd);
+		execvp(*cmd, cmd);
+		LOG(NULL, "Command", "%s failed", cmd[0]);
 		assert(0);
-	} else {
-		pid_t ws = waitpid(pid, &child_exit_status, 0);
-
-		if (ws == -1) {
-			LOG(NULL, "Failed to wait for child process", "");
-			assert(0);
-		}
-
-		if (WIFEXITED(child_exit_status)) {
-			status = WEXITSTATUS(child_exit_status);
-		} else {
-			LOG(NULL, "Child process didn't exit", "");
-			assert(0);
-		}
-
-		if (status != 0) {
-			char *cmdline = NULL;
-
-			concat_str_array(&cmdline, argv);
-			LOG(NULL, "Failed to run command:", "%s", cmdline);
-			free(cmdline);
-		}
 	}
+	return cmdpid;
+}
 
-	return status;
+void set_std_fd(int newstdinfd, int newstdoutfd, int newstderrfd)
+{
+	if (newstdinfd != -1) {
+		move_fd(newstdinfd, STDIN_FILENO);
+	}
+	if (newstdoutfd != -1) {
+		move_fd(newstdoutfd, STDOUT_FILENO);
+	}
+	if (newstderrfd != -1) {
+		move_fd(newstderrfd, STDERR_FILENO);
+	}
 }
 
-int system_argv_pipe(char *const argvp1[], int stdinp1, int stderrp1,
-		     char *const argvp2[], int stdoutp2, int stderrp2)
+void move_fd(int oldfd, int newfd)
 {
-	int statusp2;
-	int pipefd[2];
+	if (dup2(oldfd, newfd) == -1) {
```

Comment:
I would put the check in here for oldfd being != -1, and remove lines 216, 218, 219, 221, 222, and 224, but I don't feel strongly about it.

Reply:
That makes sense.  It's a good way to reduce the code duplication!